p2p/chain.py

```diff
@@ -284,13 +284,20 @@ def register_peer(self, peer: BasePeer) -> None:
             parts.extend(received)
             received_keys = set([part.unique_key for part in received])
             pending_replies -= 1
-            unexpected = received_keys.difference(
-                [key_func(header) for header in missing])
+
+            unexpected = received_keys.difference(key_func(header) for header in headers)
+            duplicates = received_keys.intersection(part.unique_key for part in parts)
```

Comment:
It looks weird to calculate duplicates here *after* adding `received` to `parts`. Seems like it will always include all the `received` keys.